includes/data-stores/class-wc-customer-download-data-store.php

```diff
@@ -61,12 +61,12 @@ public function create( &$download ) {
 			apply_filters( 'woocommerce_downloadable_file_permission_format', $format, $data )
 		);
 
-		do_action( 'woocommerce_grant_product_download_access', $data );
-
 		if ( $result ) {
 			$download->set_id( $wpdb->insert_id );
 			$download->apply_changes();
 		}
+
+		do_action( 'woocommerce_grant_product_download_access', $data );
```

Comment:
it seems the fatal error you got is caused by download ID not being set properly at that time.

The hook should be inside the `if ( $result )`. Otherwise you'd get fatal error again if `$wpdb->insert` is failed.

Reply:
Yes. Looks like so.

Let's wait for someone to confirm this though. Maybe, the issue gets triggered specifically due to some custom code in our WooCommerce installation.

Reply:
Mind sharing details on how you are hooking into this and the code you are running?

Reply:
Hello,

You should be able to see the code here: https://github.com/Automattic/woocommerce.com/pull/4662